rotkehlchen/data_import/utils.py

```diff
@@ -0,0 +1,59 @@
+from abc import ABCMeta, abstractmethod
+from pathlib import Path
+from typing import Any, List, Tuple
+
+from rotkehlchen.accounting.ledger_actions import LedgerAction
+from rotkehlchen.db.dbhandler import DBHandler
+from rotkehlchen.db.ledger_actions import DBLedgerActions
+from rotkehlchen.exchanges.data_structures import AssetMovement, Trade
+
+ROWS_PER_CONTEXT_SWITCH = 500
+
+ITEMS_PER_DB_WRITE = 200
+
+
+class BaseExchangeImporter(metaclass=ABCMeta):
+    def __init__(self, db: DBHandler) -> None:
+        self.db = db
+        self.db_ledger = DBLedgerActions(self.db, self.db.msg_aggregator)
+        self._trades: List[Trade] = []
+        self._asset_movements: List[AssetMovement] = []
+        self._ledger_actions: List[LedgerAction] = []
+
+    def import_csv(self, filepath: Path, **kwargs: Any) -> Tuple[bool, str]:
+        res = self._import_csv(filepath=filepath, **kwargs)
+        self.flush_all()  # Flush despite any errors to save processed rows
+        return res
+
+    @abstractmethod
+    def _import_csv(self, filepath: Path, **kwargs: Any) -> Tuple[bool, str]:
+        ...
+
+    def add_trade(self, trade: Trade) -> None:
+        self._trades.append(trade)
+        self.maybe_flush_all()
+
+    def add_asset_movement(self, asset_movement: AssetMovement) -> None:
+        self._asset_movements.append(asset_movement)
+        self.maybe_flush_all()
+
+    def add_ledger_action(self, ledger_action: LedgerAction) -> None:
+        self._ledger_actions.append(ledger_action)
+        self.maybe_flush_all()
+
+    def maybe_flush_all(self) -> None:
```

Comment:
this should be a private function since it's only used internally
```suggestion
    def _maybe_flush_all(self) -> None:
```

Also can you explain to me why you flush periodically? What are you trying to achieve?

I am thinking of the following problem. If for example I am importing a very big CSV and close the app in the middle, then the DB will be left in a half-populated stage.

Reply:
The reason for periodic flushes is to save RAM. Why I chose this approach:

When there was a problem with asynchronous csv import (a file was deleted after the request with `task_id` was returned) I proposed a solution to read this file and keep it in RAM but it was denied because of "something might break if the file is too big". So I thought that is this a similar situation.

P.S. before the refactoring we were writing to the database each entry one-by-one. Was it wrong then?